firmware/defmt-rtt/src/channel.rs

```diff
@@ -40,45 +44,28 @@ impl Channel {
             return 0;
         }
 
+        // calculate how much space is left in the buffer
         let read = self.read.load(Ordering::Relaxed);
         let write = self.write.load(Ordering::Acquire);
-        let available = if read > write {
-            read - write - 1
-        } else if read == 0 {
-            SIZE - write - 1
-        } else {
-            SIZE - write
-        };
-
+        let available = available_buffer_size(read, write);
+        // abort if buffer is full
         if available == 0 {
             return 0;
         }
 
-        let cursor = write;
-        let len = bytes.len().min(available);
-
-        unsafe {
-            if cursor + len > SIZE {
-                // split memcpy
-                let pivot = SIZE - cursor;
-                ptr::copy_nonoverlapping(bytes.as_ptr(), self.buffer.add(cursor), pivot);
-                ptr::copy_nonoverlapping(bytes.as_ptr().add(pivot), self.buffer, len - pivot);
-            } else {
-                // single memcpy
-                ptr::copy_nonoverlapping(bytes.as_ptr(), self.buffer.add(cursor), len);
-            }
-        }
-        self.write
-            .store(write.wrapping_add(len) % SIZE, Ordering::Release);
-
-        len
+        self.write_impl(bytes, available)
     }
 
     fn nonblocking_write(&self, bytes: &[u8]) -> usize {
-        let write = self.write.load(Ordering::Acquire);
-        let cursor = write;
-        // NOTE truncate at SIZE to avoid more than one "wrap-around" in a single `write` call
-        let len = bytes.len().min(SIZE);
+        self.write_impl(bytes, SIZE)
+    }
+
+    /// Copy `available` (or all) data from `bytes` into the buffer.
+    ///
+    /// Returns the amount of bytes copied.
+    fn write_impl(&self, bytes: &[u8], available: usize) -> usize {
+        let cursor = self.write.load(Ordering::Acquire);
```

Comment:
I get that. I've split out the not-error-prone changes to #622 and converted this one to a draft.

I'd like to keep it open, because it is a quite obvious case of code-duplication.